src/h3lib/lib/polygon.c

```diff
@@ -93,3 +95,193 @@ bool pointInsidePolygon(const GeoPolygon* geoPolygon, const BBox* bboxes,
+ * @return              Pointer to parent polygon, or null if not found
+ */
+static const LinkedGeoPolygon* findPolygonForHole(
+    const LinkedGeoLoop* loop, const LinkedGeoPolygon* polygon,
+    const BBox* bboxes, const int polygonCount) {
+    // Early exit with no polygons
+    if (polygonCount == 0) {
+        return NULL;
+    }
+    // Initialize arrays for candidate loops and their bounding boxes
+    const LinkedGeoPolygon** candidates =
+        malloc(polygonCount * sizeof(LinkedGeoPolygon*));
+    assert(candidates != NULL);
+    const BBox** candidateBBoxes = malloc(polygonCount * sizeof(BBox*));
+    assert(candidateBBoxes != NULL);
+
+    // Find all polygons that contain the loop
+    int candidateCount = 0;
+    int index = 0;
+    while (polygon) {
+        // We are guaranteed not to overlap, so just test the first point
+        if (pointInsideLinkedGeoLoop(polygon->first, &bboxes[index],
+                                     &loop->first->vertex)) {
+            candidates[candidateCount] = polygon;
+            candidateBBoxes[candidateCount] = &bboxes[index];
+            candidateCount++;
+        }
+        polygon = polygon->next;
+        index++;
+    }
+
+    // The most deeply nested container is the immediate parent
+    const LinkedGeoPolygon* parent =
+        findDeepestContainer(candidates, candidateBBoxes, candidateCount);
+
+    // Free allocated memory
+    free(candidates);
+    free(candidateBBoxes);
+
+    return parent;
+}
+
+/**
+ * Normalize a LinkedGeoPolygon in-place into a structure following GeoJSON
+ * MultiPolygon rules: Each polygon must have exactly one outer loop, which
+ * must be first in the list, followed by any holes. Holes in this algorithm
+ * are identified by winding order (holes are clockwise), which is guaranteed
+ * by the h3SetToVertexGraph algorithm.
+ *
+ * Input to this function is assumed to be a single polygon including all
+ * loops to normalize. It's assumed that a valid arrangement is possible.
+ *
+ * @param root Root polygon including all loops
+ * @return     0 on success, or an error code > 0 for invalid input
+ */
+int normalizeMultiPolygon(LinkedGeoPolygon* root) {
+    // We assume that the input is a single polygon with loops;
+    // if it has multiple polygons, don't touch it
+    if (root->next) {
+        return NORMALIZATION_ERR_MULTIPLE_POLYGONS;
+    }
+
+    // Count loops, exiting early if there's only one
+    int loopCount = countLinkedLoops(root);
+    if (loopCount <= 1) {
+        return NORMALIZATION_SUCCESS;
+    }
+
+    int resultCode = NORMALIZATION_SUCCESS;
+    LinkedGeoPolygon* polygon = NULL;
+    LinkedGeoLoop* next;
+    int innerCount = 0;
+    int outerCount = 0;
+
+    // Create an array to hold all of the inner loops. Note that
+    // this array will never be full, as there will always be fewer
+    // inner loops than outer loops.
+    LinkedGeoLoop** innerLoops = malloc(loopCount * sizeof(LinkedGeoLoop*));
```

Comment:
Why not calloc for these count * size allocations?

Reply:
Because @isaachier told me not to :). I have no particular opinion, but zeroing out the memory isn't required here.